runtime/commonMain/src/kotlinx/serialization/SerialDescriptor.kt

```diff
@@ -0,0 +1,188 @@
+ * have one element (`T`, maps have two, one for keys and one for values), but potentially unlimited
+ * number of actual children values. In such cases, valid indices range is not known statically
+ * and implementations of descriptor should provide consistent and unbounded names and indices.
+ * See [kotlinx.serialization.internal.ListLikeSerializer] for the reference.
+ *
+ * ### Thread-safety and mutability
+ * Serial descriptor implementation should be immutable after the publication and thread-safe.
+ *
+ * ### Examples
+ * TODO Roman reference to examples: color, object as list, object as map, fixed size list etc.
+ */
+public interface SerialDescriptor {
+
+    /**
+     * Serial name of the descriptor that uniquely identifies pair of the associated serializer and target class.
+     *
+     * For generated serializers, serial name is equal to the corresponding class's fully-qualified name.
+     * Custom serializers should provide a unique serial name that identify both the serializable class and
+     * the serializer itself. // TODO #363
+     */
+    public val serialName: String
+
+    /**
+     * The kind of the serialized form that determines **the shape** of the serialized data.
+     * Formats use serial kind to add and parse serializer-agnostic metadata to the result.
+     *
+     * For example, JSON format wraps [classes][StructureKind.CLASS] and [StructureKind.MAP] into
+     * brackets, while ProtoBuf just serialize these types in a separate ways.
+     *
+     * Kind should be consistent with the implementation, for example, if it is a [primitive][PrimitiveKind],
+     * then its elements count should be zero and vice versa.
+     */
+    public val kind: SerialKind
+
+    /**
+     * Whether the descriptor describes nullable element.
+     * Returns `true` if associated serializer can serialize/deserialize nullable elements of the described type.
+     */
+    public val isNullable: Boolean get() = false
+
+    /**
+     * The number of elements this descriptor describes, besides from the element itself.
+     * [elementsCount] describes the number of **semantic** elements, not the number
+     * of actual fields/properties in the serialized form, even though they frequently match.
+     *
+     * For example, for the following class
+     * `class Complex(val real: Long, val imaginary: Long)` the corresponding descriptor
+     * and the serialized form both have two elements, while for `class IntList : ArrayList<Int>()`
+     * the corresponding descriptor has a single element (`IntDescriptor`, the type of list element),
+     * but from zero up to `Int.MAX_VALUE` elements in the serialized form.
+     */
+    public val elementsCount: Int
+
+    /**
+     * Returns a serial annotations of the associated class.
+     * Serial annotations can be used to specify an additional
+     * metadata that may be used during serialization, for example a [serial name][SerialName].
```

Comment:
Actually, `@SerialName` affects code generation directly and is not stored in this array.